### PR TITLE
V3: Passing diagnostics from Rust

### PR DESCRIPTION
**Pull request by alshdavid**
Errors are currently sent to the JavaScript layer as strings. This is a problem because the JavaScript layer currently handles formatting the error output from a `Diagnostic` object - which cannot be done from a string leading to poor error messages.

This PR allows Rust to return `Diagnostic` objects to JavaScript to handle formatting and printing. This works by having Rust check if the internal `anyhow::Error` is a `Diagnostic`, if it is then it converts it to JavaScript and returns that otherwise it returns the error in string form.

On the Napi side, to serialize the `Diagnostic` value we need access to `Env` which isn't available in `deferred.reject()` so instead the error value is passed via `deferred.resolve()`. 

This then necessitated `AtlaspackNapi.buildAssetGraph()` returning a tuple and rethrowing is handled in the JavaScript layer.

```typescript
type Ok<T> = [T, null]
type Err<E> = [null, E]
type Result<T, E> = Ok<T> | Err<E>
```

## Preview

By comparison to V2 error messages, V3 error messages are pretty low on detail at this stage

#### V2 Error Message

image

#### V3 Error Message

image

## Todo

We need to add additional error information (`codeFrames`, `hints`, etc) to the error handling of our internals (resolver, transformer, etc). 

For the time being this PR allows Diagnostic errors to propage to JavaScript, following PRs should look at addressing the contents of those errors

### CODE REVIEW

File: crates/atlaspack/src/request_tracker/request_tracker.rs [hunk 1]
```diff
@@ -224,7 +224,7 @@ impl RequestTracker {
 
     *request_node = match result {
       Ok(result) => RequestNode::Valid(result.clone()),
-      Err(error) => RequestNode::Error(anyhow!(error.to_string())),
+      Err(error) => RequestNode::Error(AtlaspackError::from(error).into()),
     };
```
Comment (alshdavid) [inline crates/atlaspack/src/request_tracker/request_tracker.rs:228]: Rather then returning the error as a string, unpack the error, clone the internals and return a new error.

File: crates/atlaspack/src/request_tracker/request_tracker.rs [hunk 2]
```diff
@@ -249,7 +249,7 @@ impl RequestTracker {
     match request_node {
       RequestNode::Root => Err(diagnostic_error!("Impossible")),
       RequestNode::Incomplete => Err(diagnostic_error!("Impossible")),
-      RequestNode::Error(error) => Err(anyhow!(error.to_string())),
+      RequestNode::Error(error) => Err(AtlaspackError::from(error).into()),
```
Comment (alshdavid) [inline crates/atlaspack/src/request_tracker/request_tracker.rs:252]: And again, clone/propagate the error here

File: crates/atlaspack/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
+use anyhow::anyhow;
+use atlaspack_core::types::Diagnostic;
+use serde::Serialize;
+
+pub enum AtlaspackError {
+  Diagnostic(Diagnostic),
+  String(String),
```
Comment (mattcompiles) [inline crates/atlaspack/src/error.rs:7]: I wonder if `UnknownError` or similar would be more accurate?

**Review by mattcompiles — APPROVED** [review]
Nice one mate. Just a small question about the naming.

File: crates/atlaspack/src/error.rs (hunk 3 above)
Comment (MonicaOlejniczak) [reply crates/atlaspack/src/error.rs:7]: Agree this would be better

**Review by MonicaOlejniczak — APPROVED** [review]

**Pull request merged**